File: py/D1.py
def parseInput():
    file = open("../inputs/D1.input", "r")
    contents = file.read()
    return list(map(int, contents.splitlines()))

def p1():
    numbers = sorted(parseInput())
    L = 0
    R = len(numbers) - 1
    while (numbers[L] + numbers[R] != 2020):
        if (numbers[L] + numbers[R] < 2020):
            L += 1
        else:
            R -= 1
    return numbers[L] * numbers[R]

def p2():
    numbers = sorted(parseInput())
    for i in range(len(numbers)):
        L = i + 1
        R = len(numbers) - 1
        remainder = 2020 - numbers[i]
        while (L < R):
            if (numbers[L] + numbers[R] == remainder):
                return numbers[L] * numbers[R] * numbers[i]
            elif (numbers[L] + numbers[R] < remainder):
                L += 1
            else:
                R -= 1

print(p1())
print(p2())

File: rs/src/d1.rs
use std::fs;

pub fn p1() -> i32 {
    let mut nums = parse_input();
    nums.sort();
    let mut left = 0; 
    let mut right = nums.len() - 1;
    while nums[left] + nums[right] != 2020 {
        if nums[left] + nums[right] > 2020 {
            right -= 1;
        }
        else {
            left += 1;
        }
    }
    return nums[left] * nums[right];
}

pub fn p2() -> i32 {
    let mut nums = parse_input();
    nums.sort();
    
    for (i, n) in nums.iter().enumerate() {
        let remainder = 2020 - n;
        let mut left = i + 1;
        let mut right = nums.len() - 1;
        while left < right {
            if nums[left] + nums[right] == remainder {
                return nums[left] * nums[right] * n;
            } else if nums[left] + nums[right] > remainder {
                right -= 1;
            } else {
                left += 1;
            }
        }
    }

    return -1;
}

fn parse_input() -> Vec<i32> {
    let contents = 
        fs::read_to_string("../inputs/D1.input")
        .expect("read file");

    let mut nums = Vec::new();

    for s in contents.split("\r\n") {
        nums.push(s.parse::<i32>().expect("parse int"));
    }

    return nums;
}